Python/Product/PythonTools/PythonTools/Editor/PythonTextBufferInfo.cs

```diff
@@ -124,66 +173,89 @@ sealed class PythonTextBufferInfo : IDisposable {
         }
 
-        public event EventHandler<TextContentChangedEventArgs> OnChangedLowPriority;
-        private void Buffer_ChangedLowPriority(object sender, TextContentChangedEventArgs e) {
-            if (!_isDisposed) {
-                OnChangedLowPriority?.Invoke(this, e);
-            }
+        private void AnalysisEntry_AnalysisComplete(object sender, EventArgs e) {
+            InvokeSinks(new PythonTextBufferInfoEventArgs(PythonTextBufferInfoEvents.NewAnalysis, (AnalysisEntry)sender));
         }
 
+        private void Buffer_TextContentChanged(object sender, TextContentChangedEventArgs e) {
+            InvokeSinks(new PythonTextBufferInfoNestedEventArgs(PythonTextBufferInfoEvents.TextContentChanged, e));
+        }
+
+        private void Buffer_TextContentChangedLowPriority(object sender, TextContentChangedEventArgs e) {
+            InvokeSinks(new PythonTextBufferInfoNestedEventArgs(PythonTextBufferInfoEvents.TextContentChangedLowPriority, e));
+        }
 
-        public event EventHandler<ContentTypeChangedEventArgs> OnContentTypeChanged;
         private void Buffer_ContentTypeChanged(object sender, ContentTypeChangedEventArgs e) {
-            if (!_isDisposed) {
-                OnContentTypeChanged?.Invoke(this, e);
-            }
+            InvokeSinks(new PythonTextBufferInfoNestedEventArgs(PythonTextBufferInfoEvents.ContentTypeChanged, e));
+        }
+
+        private void Document_EncodingChanged(object sender, EncodingChangedEventArgs e) {
+            InvokeSinks(new PythonTextBufferInfoNestedEventArgs(PythonTextBufferInfoEvents.DocumentEncodingChanged, e));
         }
 
         #endregion
 
-        #region IntelliSense-related hooks
+        #region Sink management
 
-        public PythonClassifier Classifier => _classifier;
-        public PythonClassifier GetOrCreateClassifier(Func<PythonTextBufferInfo, PythonClassifier> creator) => GetOrCreate(ref _classifier, creator);
+        public void AddSink(object key, IPythonTextBufferInfoEventSink sink) {
+            if (!_eventSinks.TryAdd(key, sink)) {
+                if (_eventSinks[key] != sink) {
+                    throw new InvalidOperationException("cannot replace existing sink");
+                }
+            }
+        }
 
-        public PythonAnalysisClassifier AnalysisClassifier => _analysisClassifier;
-        public PythonAnalysisClassifier GetOrCreateAnalysisClassifier(Func<PythonTextBufferInfo, PythonAnalysisClassifier> creator) => GetOrCreate(ref _analysisClassifier, creator);
+        public T GetOrCreateSink<T>(object key, Func<PythonTextBufferInfo, T> creator) where T : class, IPythonTextBufferInfoEventSink {
+            IPythonTextBufferInfoEventSink sink;
+            if (_eventSinks.TryGetValue(key, out sink)) {
+                return sink as T;
+            }
+            sink = creator(this);
+            if (!_eventSinks.TryAdd(key, sink)) {
+                sink = _eventSinks[key];
```

Comment:
>_eventSinks [](start = 23, length = 11)

Not sure why this would find a sink when a few lines above it didn't find one. Unless the creator callback had code in it that called GetOrCreateSink?

Reply:
No telling what could be going on in the creator callback. This is the best pattern for GetOrCreate with a concurrent dictionary (except maybe disposing of IDisposable sinks if we don't use them...)

Reply:
Except their Dispose probably removes them from the buffer, and in the context of this race it will likely succeed, so they shouldn't be disposed here. (A finalizer-triggered dispose just shouldn't do the remove, since it's only for unmanaged objects.)